pkg/sourcetransformer/service.go

```diff
@@ -34,27 +41,95 @@ func (fs *Service) IsReady(context.Context, *emptypb.Empty) (*v1.ReadyResponse,
+func (fs *Service) SourceTransformFn(stream v1.SourceTransform_SourceTransformFnServer) error {
 	// handle panic
 	defer func() {
 		if r := recover(); r != nil {
 			log.Printf("panic inside sourcetransform handler: %v %v", r, string(debug.Stack()))
 			fs.shutdownCh <- struct{}{}
 		}
 	}()
-	var hd = NewHandlerDatum(d.GetValue(), d.EventTime.AsTime(), d.Watermark.AsTime(), d.Headers)
-	messageTs := fs.Transformer.Transform(ctx, d.GetKeys(), hd)
-	var results []*v1.SourceTransformResponse_Result
-	for _, m := range messageTs.Items() {
-		results = append(results, &v1.SourceTransformResponse_Result{
-			EventTime: timestamppb.New(m.EventTime()),
-			Keys:      m.Keys(),
-			Value:     m.Value(),
-			Tags:      m.Tags(),
+
+	req, err := stream.Recv()
+	if err != nil {
+		return fmt.Errorf("reading handshake message from stream: %w", err)
+	}
+
+	if req.Handshake == nil || !req.Handshake.Sot {
+		return fmt.Errorf("invalid handshake message: %+v", req)
+	}
+
+	handshakeResponse := &v1.SourceTransformResponse{
+		Handshake: &v1.Handshake{
+			Sot: true,
+		},
+	}
+	if err := stream.Send(handshakeResponse); err != nil {
+		return fmt.Errorf("sending handshake response to client over gRPC stream: %w", err)
+	}
+
+	ctx := stream.Context()
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	grp, grpCtx := errgroup.WithContext(ctx)
+
+	senderCh := make(chan *v1.SourceTransformResponse, 500) // TODO: identify the right buffer size
+	// goroutine to send the response to the stream
+	grp.Go(func() error {
+		for {
+			select {
+			case <-grpCtx.Done():
+				return grpCtx.Err()
+			default:
+			}
+			if err := stream.Send(<-senderCh); err != nil {
+				cancel()
+				return err
+			}
+		}
+	})
+
+	for {
+		d, err := stream.Recv()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
+			return err
+		}
+		if d.Handshake != nil {
+			return fmt.Errorf("expected source transform messages, received handshake message")
+		}
+
+		req := d.Request
+		grp.Go(func() error {
+			var hd = NewHandlerDatum(req.GetValue(), req.EventTime.AsTime(), req.Watermark.AsTime(), req.Headers)
+			messageTs := fs.Transformer.Transform(grpCtx, req.GetKeys(), hd)
+			var results []*v1.SourceTransformResponse_Result
+			for _, m := range messageTs.Items() {
+				results = append(results, &v1.SourceTransformResponse_Result{
+					EventTime: timestamppb.New(m.EventTime()),
+					Keys:      m.Keys(),
+					Value:     m.Value(),
+					Tags:      m.Tags(),
+				})
+			}
+			resp := &v1.SourceTransformResponse{
+				Results: results,
+				Id:      req.GetId(),
+			}
+			select {
+			case senderCh <- resp:
+			case <-grpCtx.Done():
+				return grpCtx.Err()
+			}
+			return nil
 		})
 	}
-	responseList := &v1.SourceTransformResponse{
-		Results: results,
+
+	if err := grp.Wait(); err != nil {
+		statusErr := status.Errorf(codes.Internal, err.Error())
+		return statusErr
 	}
-	return responseList, nil
+	return nil
```

Comment:
if there is an error we will not let the client know until the stream is closed?